### PR TITLE
Disable tracking memory allocation counts by default

### PR DESCRIPTION
**Pull request by slouken**
Fixes https://github.com/libsdl-org/SDL/issues/11099

### CODE REVIEW

**Comment by slouken** [comment]
One side effect of this change is that we might get false positives from free() validation when SDLTest_TrackAllocations() is used. I think this is okay, but pinging @madebr and @icculus for confirmation.

**Comment by madebr** [comment]
> One side effect of this change is that we might get false positives from free() validation when SDLTest_TrackAllocations() is used. I think this is okay, but pinging @madebr and @icculus for confirmation.

I don't think so. SDL_test wraps the original allocation functions with its own using `SDL_SetMemoryFunctions`.

While you're working in SDL_malloc.c, shouldn't `SDL_SetMemoryFunctions` and `SDL_GetMemoryFunctions` be atomic functions? On the other hand, if you're calling these functions in multiple threads, then you have bigger issues.

**Comment by slouken** [comment]
> > One side effect of this change is that we might get false positives from free() validation when SDLTest_TrackAllocations() is used. I think this is okay, but pinging @madebr and @icculus for confirmation.
> 
> I don't think so? SDL_test wraps the original allocation functions with its own using `SDL_SetMemoryFunctions`.

See the emscripten CI failure.

> On the other hand, if you're calling these functions in multiple threads, then you have bigger issues.

Agreed.

**Comment by madebr** [comment]
> > > One side effect of this change is that we might get false positives from free() validation when SDLTest_TrackAllocations() is used. I think this is okay, but pinging @madebr and @icculus for confirmation.
> > 
> > 
> > I don't think so? SDL_test wraps the original allocation functions with its own using `SDL_SetMemoryFunctions`.
> 
> See the emscripten CI failure.

I see.
Emscripten fails while destroying the environment mutex:
```
$SDL_ReportAssertion	@	SDL_assert.c:374
$SDLTest_TrackedFree	@	SDL_test_memory.c:281
$SDL_free	@	SDL_malloc.c:6495
$SDL_DestroyMutex	@	SDL_sysmutex.c:60
$SDL_DestroyEnvironment	@	SDL_getenv.c:296
$SDL_QuitEnvironment	@	SDL_getenv.c:80
$SDL_QuitMainThread	@	SDL.c:301
$SDL_Quit	@	SDL.c:669
$SDL_main	@	testbounds.c:53
$SDL_RunApp	@	SDL_sysmain_runapp.c:53
$main	@	SDL_main_impl.h:137
```

**Comment by madebr** [comment]
hould we make SDL_test's memory tracking more flexible and assume non-tracked memory is allocated by the original allocator?

**Comment by slouken** [comment]
> Should we make SDL_test's memory tracking more lenient and assume non-tracked memory is allocated by the original allocator?

Yes, this seems like a good idea. Feel free to push to my branch.

**Comment by madebr** [comment]
`SDL_GetNumAllocations()`'s documentation should mention that it returns `0` by default, unless SDL is built with allocation tracking support (`-DTRACK_ALLOCATION_COUNT`).

File: src/stdlib/SDL_malloc.c [hunk 1]
```diff
@@ -6352,6 +6352,17 @@ static struct
     real_malloc, real_calloc, real_realloc, real_free, { 0 }
 };
 
+// Define this if you want to track the number of allocations active
+// #define TRACK_ALLOCATION_COUNT
+#ifdef TRACK_ALLOCATION_COUNT
```
Comment (madebr) [inline src/stdlib/SDL_malloc.c:6357]: Should the macro start with `SDL_`?

**Comment by madebr** [comment]
Is returning `-1` when allocation counting is disabled a breaking change?

<details>

<summary>Proposed change</summary>

```diff
diff --git a/include/SDL3/SDL_stdinc.h b/include/SDL3/SDL_stdinc.h
index d9ef590c5..edbee6b41 100644
--- a/include/SDL3/SDL_stdinc.h
+++ b/include/SDL3/SDL_stdinc.h
@@ -1024,7 +1024,8 @@ extern SDL_DECLSPEC void SDLCALL SDL_aligned_free(void *mem);
 /**
  * Get the number of outstanding (unfreed) allocations.
  *
- * \returns the number of allocations.
+ * \returns the number of allocations or
+ *          -1 if allocation counting is disabled.
  *
  * \threadsafety It is safe to call this function from any thread.
  *
diff --git a/src/stdlib/SDL_malloc.c b/src/stdlib/SDL_malloc.c
index 72e5354c9..c1c1fa1f0 100644
--- a/src/stdlib/SDL_malloc.c
+++ b/src/stdlib/SDL_malloc.c
@@ -6428,7 +6428,11 @@ bool SDL_SetMemoryFunctions(SDL_malloc_func malloc_func,
 
 int SDL_GetNumAllocations(void)
 {
+#ifdef TRACK_ALLOCATION_COUNT
     return SDL_GetAtomicInt(&s_mem.num_allocations);
+#else
+    return -1;
+#endif
 }
 
 void *SDL_malloc(size_t size)
```

</details>

File: src/stdlib/SDL_malloc.c (hunk 1 above)
Comment (slouken) [reply src/stdlib/SDL_malloc.c:6357]: Yes, since we’re documenting this that’s a good idea.

**Comment by slouken** [comment]
> Is returning `-1` when allocation counting is disabled a breaking change?
> 
> Proposed change

That seems reasonable, and a good way to detect whether the API is returning valid results.

**Comment by slouken** [comment]
I’m out until later, so feel free to push and/or merge. I’ll take care of whatever’s left when I get back.

**Comment by madebr** [comment]
> I’m out until later, so feel free to push and/or merge. I’ll take care of whatever’s left when I get back.

Done. I leave the squashing and spell checking to you :)

Happy Christmas shopping ;)

**Comment by slouken** [comment]
Merged, thanks!

File: include/SDL3/SDL_stdinc.h [hunk 2]
```diff
@@ -1024,7 +1024,8 @@ extern SDL_DECLSPEC void SDLCALL SDL_aligned_free(void *mem);
 /**
  * Get the number of outstanding (unfreed) allocations.
  *
- * \returns the number of allocations.
+ * \returns the number of allocations or
+ *          -1 if allocation counting is disabled.
```
Comment (madebr) [inline include/SDL3/SDL_stdinc.h:1028]: I wrote -1 here, should we generalize it to negative values?

Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1028]: No, -1 is correct and fine.

**Pull request merged**